### ISSUE TITLE
Pnpm seems to be consistently slower than yarn (classic)

### ISSUE THREAD

**Issue by meeroslav** [pnpm/pnpm#6447]
At [nx](https://github.com/nrwl/nx), we test our tools with major package managers - npm, yarn and pnpm.

Our [nightly tests](https://github.com/nrwl/nx/actions/runs/4759975498) are running a matrix and recently we added reporting of the duration of each package run. The result were surprising:

![Screenshot 2023-04-21 at 17 06 02](https://user-images.githubusercontent.com/881612/233670317-07cb4fbf-8404-437d-bfdd-bffd0f7809f7.png)

We have a test branch where our repo is migrated to `pnpm`, but that branch gives similar results where `pnpm` keeps running significantly slower.

I did some troubleshooting today running `SELECTED_PM=pnpm yarn nx e2e e2e-rollup` on our repo which does the following:
- generates a new repo with `create-nx-workspace`
- generates a library and installs necessary dependencies
- runs builds

![Screenshot 2023-04-21 at 17 09 14](https://user-images.githubusercontent.com/881612/233671473-5db8146d-fc57-4c2f-814d-404904b08a17.png)

The `newProject` is just:
- creating an empty temp folder with `package.json`
- runs `yarn/pnpm add -D` for a group of packages
- moves that folder to a new destination

![image](https://user-images.githubusercontent.com/881612/233672176-5d6dc71e-8fa9-4766-9c5e-f12fef1bcc2c.png)

Is there something crucial we are missing?

We even tested building typical `create-react-app`, removing `node_modules` and running `yarn --frozen-lockfile` vs `pnpm i --frozen-lockfile` and yarn is consistently faster.

Please note that there are no special flags in `.npmrc`.

### Expected behavior:

I would expect `pnpm` to be faster or at least comparable in speed

### Actual behavior:

The `pnpm` is consistently slower, even for running custom commands e.g. `yarn nx ...` vs `pnpm exec nx ...` 

### Additional information:

 - `node -v` prints: v18.12.1
 - Windows, macOS, or Linux?: Nightly runs on Ubuntu, Locally reproducible on Mac M1

**Comment by RexSkz** [comment]
I reproduced it on my Windows. It's an example that is quite simple, with only two files in a repo and no need to install dependencies:

```json
{
  "name": "pnpm-test",
  "version": "1.0.0",
  "scripts": {
    "echo": "echo 'hello world'",
    "bench:npm": "sh ./bench.sh npm",
    "bench:yarn": "sh ./bench.sh yarn",
    "bench:pnpm": "sh ./bench.sh pnpm"
  }
}
```

```bash
#!/bin/sh

case $1 in
    "npm")
        for i in {1..10}; do npm run echo; done
        ;;
    "yarn")
        for i in {1..10}; do yarn run echo; done
        ;;
    "pnpm")
        for i in {1..10}; do pnpm run echo; done
        ;;
    *)
        echo "Usage: bench-npm.sh [npm|yarn|pnpm]"
        ;;
esac
```

```bash
# time yarn bench:npm
real    0m5.761s
user    0m0.030s
sys     0m0.091s

# time yarn bench:yarn
real    0m3.926s
user    0m0.000s
sys     0m0.000s

# time yarn bench:pnpm
real    0m5.992s
user    0m0.000s
sys     0m0.030s
```

Considering there are no actual works, is it because PNPM performs some checks or actions before running a script?

**Comment by zkochan** [comment]
`pnpm run` commands are faster if you install pnpm using the [standalone script](https://pnpm.io/installation#using-a-standalone-script) or you install the `@pnpm/exe` package, not the pnpm package from npmjs.

**Comment by meeroslav** [comment]
Thank you, @zkochan,

I was following the guides on https://pnpm.io/continuous-integration, assuming this is the optimal setup. They suggest using `corepack` (for CircleCI) and `pnpm/action-setup` (for Github).

I'll try installing the standalone script on CI and will get back with the results.

**Comment by meeroslav** [comment]
I reran my tests. Despite seeing some improvement after removing the `pnpm` and installing the standalone version, it's still behind `yarn` (the `pnpm2` is the standalone version).

![Screenshot 2023-04-24 at 11 41 36](https://user-images.githubusercontent.com/881612/233960254-6ca50e0b-8d5b-4a14-8d6f-65ad3a12d7a2.png)

I also ran the test by @RexSkz with standalone `pnpm` and used `pnpm` as a test runner, and there I finally saw the improvement:
![Screenshot 2023-04-24 at 11 45 08](https://user-images.githubusercontent.com/881612/233961158-b829bcc4-3292-4fa8-a9f6-94d8f7535b66.png)
![Screenshot 2023-04-24 at 11 44 56](https://user-images.githubusercontent.com/881612/233961100-a13913c5-e995-44a5-952e-b461ce0574a6.png)
![Screenshot 2023-04-24 at 11 45 21](https://user-images.githubusercontent.com/881612/233961202-d450781e-2afc-4299-890a-4804abf51130.png)

Interestingly, selecting a different package manager as a main runner also makes a tremendous difference:
![Screenshot 2023-04-24 at 11 48 15](https://user-images.githubusercontent.com/881612/233961934-46242c9f-23e7-423b-8467-d329294613b4.png)

**Comment by meeroslav** [comment]
I made another test where I ran the simple command `nx --version` (with `npx/yarn/pnpm`) in all combinations:

(in the repo with `pnpm-lock.yaml`)
![Screenshot 2023-04-24 at 12 01 02](https://user-images.githubusercontent.com/881612/233964872-f872b27f-4c1e-4b10-8a79-7d0713620017.png)

(in the repo with `package-lock.json`)
![Screenshot 2023-04-24 at 12 08 45](https://user-images.githubusercontent.com/881612/233966588-2a10aa6d-b999-4ff2-8c5e-6e72e92e8a2e.png)

(in the repo with `yarn.lock`)
![Screenshot 2023-04-24 at 12 11 51](https://user-images.githubusercontent.com/881612/233967218-08936491-6c13-4362-9008-facaa6b4de98.png)

Running commands with standalone `pnpm` seems to be the fastest, BUT running the script that contains the `pnpm` command is fastest when run with `npm`.

I did not expect such turn of events.

**Comment by RexSkz** [comment]
> Running commands with standalone pnpm seems to be the fastest, BUT running the script that contains the pnpm command is fastest when run with npm.

It may be due to the Node.js bytecode or module cache, but I have no evidence.

Also, I'm curious why `@pnpm/exe` runs faster. Is it because `@pnpm/exe` has its own Node.js that makes it load fewer packages when executed?

**Comment by meeroslav** [comment]
It would be interesting to get that info. Currently, `@pnpm/exe` is comparable (if not faster) to `standalone`.

Standalone:
![Screenshot 2023-04-24 at 12 43 07](https://user-images.githubusercontent.com/881612/233973865-54f37834-d532-423a-9e8d-6a0934abde53.png)

With `npm i -g @pnpm/exe`:
![Screenshot 2023-04-24 at 12 45 23](https://user-images.githubusercontent.com/881612/233974302-56e01516-db1f-441e-8a45-b5448935e3c5.png)

Standard `npm i -g pnpm`
![Screenshot 2023-04-24 at 12 46 35](https://user-images.githubusercontent.com/881612/233974574-0b984477-b0ae-46b8-bb1d-5d903cb0daff.png)

**Comment by meeroslav** [comment]
What would the reason for not advertising `@pnpm/exe` as a default means of installation? 

Are there any downsides I should know of?

**Comment by zkochan** [comment]
It is advertised as the default mean of installation. The standalone scripts are the first on the installation page.

On CI it might make less sense because you need to install pnpm on each run and installing `@pnpm/exe` is slower then installing the plain js version.

**Comment by meeroslav** [comment]
Using `@pnpm/exe` on CI did give us a little bit of a boost. Unfortunately, not enough to catch up with `yarn`.

Using standard `pnpm/action-setup`:
```
| PM   | Total time  |
|------|-------------|
| yarn | 4h 35m 25s  |
| pnpm | 6h 9m 13s   |
```

Using `npm i -g @pnpm/exe`:
```
| PM   | Total time  |
|------|-------------|
| yarn | 4h 34m 57s  |
| pnpm | 5h 44m 7s   |
```

In some cases, these numbers are quite drastic:
```
| Project                        | Time                      |
|--------------------------------|---------------------------|
| e2e-nx-run                     |                           |
|                         min ❗ | 14m 13s (Linux, yarn)     |
|                         max ❌ | 26m 47s (Linux, pnpm)     |
| e2e-angular-core               |                           |
|                         min ❌ | 17m 8s (Linux, yarn)      |
|                         max ❌ | 23m 5s (Linux, pnpm)      |
| e2e-esbuild                    |                           |
|                         min ✅ | 6m 8s (Linux, yarn)       |
|                         max ✅ | 10m 7s (Linux, pnpm)      |
| e2e-lerna-smoke-tests          |                           |
|                         min ✅ | 2m 16s (Linux, yarn)      |
|                         max ✅ | 5m 15s (Linux, pnpm)      |
| e2e-next                       |                           |
|                         min ❌ | 20m 58s (Linux, yarn)     |
|                         max ❌ | 28m 47s (Linux, pnpm)     |
| e2e-node                       |                           |
|                         min ❌ | 16m 32s (Linux, yarn)     |
|                         max ❌ | 27m 53s (Linux, pnpm)     |
| e2e-storybook                  |                           |
|                         min ✅ | 4m 24s (Linux, yarn)      |
|                         max ✅ | 9m 58s (Linux, pnpm)      |
| e2e-web                        |                           |
|                         min ❗ | 14m 12s (Linux, yarn)     |
|                         max ❌ | 21m 4s (Linux, pnpm)      |
```

Please let me know if you need an example scenario to investigate and debug this further.

**Comment by RexSkz** [comment]
I tried to figure out why PNPM starts slower than Yarn. Here are some conclusions.

TL;DR: it's due to the large number of modules imported when PNPM starts. There IS a solution, but it can't be done quickly.

---

## Using Flame Graph

At first, I was to figure out what PNPM does when performing `pnpm run ...`. I used the built-in `--prof` in Node.js to collect and transform data:

```bash
# In a project with the script "echo" in package.json
node --prof ../pnpm/pnpm/lib/pnpm.js echo
node --prof-process --preprocess -j isolate-*.log > pnpm.json
```

Open it with [flamebearer](https://mapbox.github.io/flamebearer/), and here's the flame graph (the green parts are the functions in PNPM package):

<img width="782" alt="image" src="https://user-images.githubusercontent.com/27483702/234520018-6c951048-1cdf-4cce-a889-7edf7e71fea3.png">

It looks normal, and there's no time-consuming operation, but some packages interested me, like `plugin-commands-deploy` and `tarball-fetcher`. Does PNPM really need these packages each time it starts?

But when I set breakpoints in their main functions, no break at all, which means they're not executed. That's weird.

## Using Breakpoints

I set breakpoints at both `pnpm/lib/main.js:main` and `exec/plugin-commands-script-runners/lib/run.js:runScript`, but I didn't find too much time elapsed between them, which means once the `main` function is called, PNPM will finish its job quickly.

But if I set the breakpoints at both `pnpm/lib/main.js:1` and `pnpm/lib/main.js:main`, I found a 1-second gap between them, which means the time is spent on `import` statements in `main.js`.

## Time Spent in `import`

Using some simple code, I can figure out how much time each `import` statement costs:

![image](https://user-images.githubusercontent.com/27483702/234525727-1a80ee38-b71a-4bb2-af53-3fae06ab9e34.png)

If you keep tracing, the call graph looks just like the flame graph before! This means there is no problem with the PNPM functions; it's because PNPM has too many modules to import, which causes it starts slowly.

I tried to comment out some imports that will not be used in `pnpm run`, and the executed time decreased from 1300 ms to 300 ms:

![image](https://user-images.githubusercontent.com/27483702/234527052-f6e31d94-dd7e-40d1-baa4-2d0ca4551ef0.png)

## Why Yarn is Faster

I used [import-graph-visualizer](https://github.com/rx-angular/import-graph-visualizer) to visualize the dependency in PNPM & Yarn repo.

The result of `import-graph-visualizer --entry-points pnpm/src/pnpm.ts`:

<img width="616" alt="image" src="https://user-images.githubusercontent.com/27483702/234527743-b34a3cf8-54ba-46a0-9bba-d07d15b0a6cb.png">

The result of `import-graph-visualizer --entry-points packages/yarnpkg-cli/sources/cli.ts`:

<img width="424" alt="image" src="https://user-images.githubusercontent.com/27483702/234527877-a1622463-593e-45b0-9e0d-7586a3ccb193.png">

Though it may not be accurate, it can tell that Yarn has fewer imports than PNPM.

## The Solution

- Use dynamic `import` instead of top-level `import`, which has been applied in the file `pnpm/src/pnpm.ts`, but it's not enough.
- Separate some constants/utils into smaller files to prevent always importing the whole module.

Unfortunately, it's huge work and can't be done quickly.

**Comment by zkochan** [comment]
@RexSkz I don't understand this. Both Yarn and pnpm are bundled into a single js file. How do dynamic imports then work in Yarn?

**Comment by RexSkz** [comment]
@zkochan Whether bundling them into a single file doesn't change top-level and dynamic import behaviour too much. On my laptop (M1 Mac), PNPM and Yarn have a similar compilation time (~120ms), but PNPM has much more execution time.

Here's the flame graph of the CPU profile of the command `node --cpu-prof ~/.nvm/versions/node/v16.19.0/lib/node_modules/pnpm/bin/pnpm.cjs echo`:

<img width="1639" alt="image" src="https://user-images.githubusercontent.com/27483702/234778776-3e802b61-22a4-40ba-b9d0-2fd120ed97cf.png">

You can see the shape of "the 'require' hell" part looks almost the same as the development version shown below:

<img width="1575" alt="image" src="https://user-images.githubusercontent.com/27483702/234779761-bfaee2c6-2797-463a-95da-04e611d1f6d9.png">

**The reason is that both top-level and dynamic imports will be transformed to `require`. Node.js will execute the top-level codes when it requires a module for the first time. Using top-level import will lead to the import hell before the function is executed, while using dynamic import will let Node.js execute the codes only when necessary.**

As a comparison, here's the flame graph of the CPU profile of the command `node --cpu-prof ~/.yarn/lib/cli.js echo`:

<img width="1633" alt="image" src="https://user-images.githubusercontent.com/27483702/234781488-8fc80f3a-860f-4ffb-a8c9-8159d4e4c4d8.png">

---

I understand that PNPM has its own design, and changing the top-level import to dynamic import is not that easy. It needs help from the dependency analytic tools, and sometimes we must change a file's structure or a function's logic.

Using other PMs to execute a script also costs extra time (for Yarn it's ~160ms each time on my laptop), so I think maybe we should not use PMs to execute scripts if they'll be executed so many times.

BTW it's the first time I realize that import will harm the performance...😂

**Comment by zkochan** [comment]
If we can improve performance by using dynamic requires then we should try to do that. I think actually there is a babel plugin to automatically convert all requires to dynamic ones. This is used by bit CLI (cc @davidfirst and @GiladShoham)

**Comment by davidfirst** [comment]
Just be careful if you use this plugin to keep an eye on circular dependencies. Because it automatically converts all import statements to dynamically require, it'll keep the code running no matter how many circulars you have, which otherwise would fail. So maybe pair this with an eslint rule that checks for circulars.

**Comment by JacobLey** [comment]
If the issue is largely related to parsing/loading of modules, this could also be mitigated by migrating to ESM?

ESM performs all top-level `import`s asynchronously and in parallel.

That may mitigate the issue without introducing a new builder tool?

The biggest pushback against ESM is usually that it breaks for _consumers_ of the package using commonjs.
However to my knowledge pnpm doesn't advertise much of a programmatic API so that should be a non-issue?

The code is already written as ESM (using `import`/`export` syntax) so ideally adoptions would mostly just require tweaks to tsconfig (`module`) and package.json (`type`).

The biggest issue would be local path imports would need to be updated to have `.js` extension. PITA admittedly, but straightforward

**Comment by RexSkz** [comment]
> If the issue is largely related to parsing/loading of modules, this could also be mitigated by migrating to ESM?
> 
> ESM performs all top-level `import`s asynchronously and in parallel.
> 
> That may mitigate the issue without introducing a new builder tool?
> 
> The biggest pushback against ESM is usually that it breaks for _consumers_ of the package using commonjs. However to my knowledge pnpm doesn't advertise much of a programmatic API so that should be a non-issue?
> 
> The code is already written as ESM (using `import`/`export` syntax) so ideally adoptions would mostly just require tweaks to tsconfig (`module`) and package.json (`type`).
> 
> The biggest issue would be local path imports would need to be updated to have `.js` extension. PITA admittedly, but straightforward

Switching to ESM can't reduce the loading time because all top-level codes are executed at the beginning (by the main thread), only dynamic import can make sure the codes are executed ONLY if the module is imported.

**Comment by JacobLey** [comment]
> all top-level codes are executed
Agreed, ESM will do this loading/excution faster than it's raw CommonJS equivalent, but isn't the same as `lazy` which fully defers this execution to the future (which depending on use case, is never! Hard to beat 0!).
But if it does eventually get loaded in main thread, it still gets hit with long blocking synchronous loading.

Conceptually could _also_ implement lazy with ESM, but not sure how trivial that is. IDK if `babel` supports that (tool used in current draft PR). SWC _claims_ to support it, but have not tested.

Given the async nature of `import` it would only be usable in async contexts... admittedly that is most though.

Are there any good real world benchmarks of ESM async import vs CommonJS lazy import?

If the current pnpm CI includes some performance tracking
> I don't know what change caused it but I have noticed that in our benchmarks

(sounds like it does?)

I would be open to taking a swing at a counter PR implementing ESM? And then can compare stats more directly?
Could be a good case study for future projects as well for whether ESM has real world performance gains (rather than just being the "future syntax" of JS)

**Comment by JacobLey** [comment]
Looked into this a bit more, and realized that the 
> Both Yarn and pnpm are bundled into a single js file

aspect would probably make ESM adoption a non-starter. At least for the reasoning of performance (may still be preferable for other reasons...)

First off most bundlers don't handle ESM super well. Most expect to transpile to CJS+require. Even if they did successfully bundle the code, it is a single large file. _Every_ line gets parsed before execution, regardless of what parts _actually_ dynamically run. So the part where ESM will get the best performance gains is immediately wiped out.

FWIW, I expect lazy commonjs will _also_ have a similar issue. It will probably run faster than it's non-lazy counterpart, but not to the degree desired.

Based on RexSkz's charts above (if I am reading them right), the mere _compiling_/_loading_ of the bundled file (which is over 200k lines long BTW!) is ~125ms. Before require-hell even starts.

Yarn's equivalent is done compiling at ~95ms, and is done with its main requires by ~130ms.

So pnpm would have to cut down it's actual lazy execution to _only_ 5ms to be successfully competing with yarn. I don't think lazy loading is going to get that win...

My next thought is why is pnpm bundling into a single file anyways?
I suspect it is for simplicity (don't need to ship full `node_modules` or dependencies), but I think the size of this executable is massively hurting performance.

What about using `pnpm`'s own `deploy` functionality to create a specialized package?
Then introducing ESM could potentially have a much larger impact.

**Comment by SukkaW** [comment]
> Here's the flame graph of the CPU profile of the command node --cpu-prof ~/.nvm/versions/node/v16.19.0/lib/node_modules/pnpm/bin/pnpm.cjs echo:

@RexSkz 

I have noticed something else from the flamegraph.

It only takes yarn 94ms before entering the "require hell" (`__webpack_require__` hell to be exact). But for pnpm, it needs 125ms before entering the "require-hell".

Both yarn and pnpm ship a big chunky file, but the size of pnpm's `pnpm.cjs` is 8.4 MiB while the size of yarn's `cli.js` is at 5.31 MiB. So in theory, if we can reduce the size of `pnpm.cjs`, we can reduce the cold start time of pnpm.

**Comment by SukkaW** [comment]
Tested `v8-compile-cache + pnpm` with this setup:

```bash
$ tree
.
├── bench.sh
├── node_modules
├── p
├── package.json
├── pnpm-dist # run `pnpm run compile` inside pnpm projects, then copy the `pnpm/dist` here
└── pnpm-lock.yaml

$ cat package.json
{
  "name": "pnpm-test",
  "version": "1.0.0",
  "scripts": {
    "echo": "echo 'hello world'"
  },
  "dependencies": {
    "v8-compile-cache": "^2.4.0"
  }
}

$ cat bench.sh
#!/bin/sh

case $1 in
    "npm")
        for i in {1..10}; do npm run echo > /dev/null; done
        ;;
    "yarn")
        for i in {1..10}; do yarn run echo > /dev/null; done
        ;;
    "pnpm")
        for i in {1..10}; do pnpm run echo > /dev/null; done
        ;;
    "p-v8")
        for i in {1..10}; do node ./p-v8 run echo > /dev/null; done
        ;;
    *)
        echo "Usage: bench-npm.sh [npm|yarn|pnpm]"
        ;;
esac

$ cat ./p-v8
#!/usr/bin/env node
require('v8-compile-cache'); // here we require v8-compile-cache
require('./pnpm-dist/pnpm.cjs')
```

The cold boot time of pnpm:

```bash
$ time ./bench.sh pnpm
./bench.sh pnpm  4.51s user 0.95s system 101% cpu 5.377 total
```

The cold boot time of `v8-compile-cache` + `pnpm.cjs`:

```bash
$ time ./bench.sh p
./bench.sh p  4.07s user 0.67s system 102% cpu 4.613 total
```

With `v8-compile-cache` pnpm can accomplish a 14% improvement in cold boot time.

**Comment by RexSkz** [comment]
@SukkaW It's weird that there's no `internalCompileFunction` on my second flamegraph; maybe Node.js has its own compile cache? (I'm not sure)

Anyway, using `v8-compile-cache` seems to be *the easiest way to improve some performance*, and it seems safe. If Yarn uses it, PNPM can give it a try as well.

**Comment by SukkaW** [comment]
> If we can improve performance by using dynamic requires then we should try to do that. I think actually there is a babel plugin to automatically convert all requires to dynamic ones.

@zkochan @RexSkz I've created a esbuild Feature Request (https://github.com/evanw/esbuild/issues/3488). If esbuild implement this, we could solve the "require hell" overhead with a single build option.

**Comment by sysmat** [comment]
whit each new version of pnpm it get slower & slower

**Comment by SukkaW** [comment]
> I've created a esbuild Feature Request ([evanw/esbuild#3488](https://github.com/evanw/esbuild/issues/3488)). If esbuild implements this, we could solve the "require hell" overhead with a single build option.

Update: esbuild refuses to implement the lazy require transformation, but we can still do it by implementing an esbuild plugin.

@zkochan Should we investigate this further?

**Comment by zkochan** [comment]
> The biggest pushback against ESM is usually that it breaks for consumers of the package using commonjs.
However to my knowledge pnpm doesn't advertise much of a programmatic API so that should be a non-issue?

pnpm's programmatic API is used by [Bit](https://github.com/teambit/bit), who are also the biggest sponsors of pnpm at the moment. But we could probably ship both and esm and non-esm version of pnpm, I guess.

> Looked into this a bit more, and realized that the
>>Both Yarn and pnpm are bundled into a single js file
>
>aspect would probably make ESM adoption a non-starter. At least for the reasoning of performance (may still be preferable for other reasons...)

we can ship pnpm in multipe js files if it will make it faster. That's not a problem.

> whit each new version of pnpm it get slower & slower

Not according to the benchmarks. I don't see any changes.

> Should we investigate this further?

You are welcome to investigate this. I have other priorities at the moment.

**Comment by andrew-s-faulkner** [comment]
This is still happening - some rough benchmarks on a 2023 Mac OS with an M3 chip and 36gb of RAM:

| Test                             | npm (s) | pnpm (s) | Result                 | OS  | Node version |
|----------------------------------|---------|----------|------------------------|------|-----------|
| Initial install                  | 36.70   | 13.66    | pnpm: 2.69x faster     | Mac  | 22.9.0    |
| Initial install                  | 38.64   | 17.24    | pnpm: 2.24x faster     | Mac  | 22.9.0    |
| Initial install [run 1]          | 17.46   | 28.26    | npm: 1.62x faster      | Mac  | 20.15.1   |
| Initial install [run 2]          | 20.16   | 18.71    | pnpm: 1.08x faster     | Mac  | 20.15.1   |
| With lockfile [run 1]            | 11.49   | 11.57    | Tie [effectively]                    | Mac  | 20.15.1   |
| With lockfile [run 2]            | 11.79   | 11.61    | Tie [effectively]                   | Mac  | 20.15.1   |
| With lockfile [run 3]            | 11.57   | 12.22    | pnpm: 1.06x faster     | Mac  | 20.15.1   |
| With lockfile                    | 6.38    | 9.84     | npm: 1.54x faster      | Mac  | 22.9.0    |
| With lockfile                    | 4.83    | 7.77     | npm: 1.61x faster      | Mac  | 22.9.0    |
| With lockfile                    | 7       | 1.8      | pnpm: 3.89x faster       | WSL  | 20.15.1   |
| Single package install ([async](https://www.npmjs.com/package/async) package [*]) | 1.29    | 1.96     | npm: 1.52x faster      | Mac  | 20.15.1   |
| Single package install ([async](https://www.npmjs.com/package/async) package [*]) | 1.44    | 1.8      | npm: 1.25x faster      | Mac  | 20.15.1   |

[*] literally `[p]npm install async` (as in the module)

**Comment by zkochan** [comment]
This no longer reproduces on current pnpm, so closing.

pnpm is a native binary now and installs are several times faster than the version this was measured on.

If you still hit this on the latest pnpm, please comment with a reproduction and we will reopen.

---
Written by an agent (Claude Code, claude-opus-5).
